File: back-kotlin/src/main/kotlin/me/nicolas/stravastats/domain/business/badges/BadgeSet.kt
package me.nicolas.stravastats.domain.business.badges

import me.nicolas.stravastats.domain.business.strava.StravaActivity


enum class BadgeSetEnum {
    FAMOUS, GENERAL
}

/**
 * A BadgeSet is a collection of badges.
 * @param name the name of the badge set
 * @param badges the list of badges
 */
data class BadgeSet(val name: String, private val badges: List<Badge> = listOf()) {

    /**
     * Check all the badges of the set.
     * @param activities the list of activities to check
     * @return a list of BadgeCheckResult
     */
    fun check(activities: List<StravaActivity>): List<BadgeCheckResult> {
        val results = badges.map { badge ->
            val (checkedActivities, isCompleted) = badge.check(activities)
            BadgeCheckResult(badge, checkedActivities, isCompleted)
        }
        return deduplicateFamousClimbActivities(results)
    }

    operator fun plus(anotherBadgeSet: BadgeSet): BadgeSet {
        return BadgeSet(name = name, badges = badges + anotherBadgeSet.badges)
    }

    private data class FamousClimbGroupKey(
        val name: String,
        val latitude: Double,
        val longitude: Double,
    )

    private data class FamousClimbActivityKey(
        val group: FamousClimbGroupKey,
        val activityId: Long,
    )

    private data class FamousClimbActivityWinner(
        val resultIndex: Int,
        val quality: Double,
    )

    private fun deduplicateFamousClimbActivities(results: List<BadgeCheckResult>): List<BadgeCheckResult> {
        val winners = mutableMapOf<FamousClimbActivityKey, FamousClimbActivityWinner>()
        results.forEachIndexed { resultIndex, result ->
            val badge = result.badge as? FamousClimbBadge ?: return@forEachIndexed
            val group = FamousClimbGroupKey(badge.name.trim().lowercase(), badge.end.latitude, badge.end.longitude)
            result.activities.forEach { activity ->
                val quality = badge.matchQuality(activity) ?: return@forEach
                val key = FamousClimbActivityKey(group, activity.id)
                val winner = winners[key]
                if (winner == null || quality < winner.quality) {
                    winners[key] = FamousClimbActivityWinner(resultIndex, quality)
                }
            }
        }

        return results.mapIndexed { resultIndex, result ->
            val badge = result.badge as? FamousClimbBadge ?: return@mapIndexed result
            val group = FamousClimbGroupKey(badge.name.trim().lowercase(), badge.end.latitude, badge.end.longitude)
            val filteredActivities = result.activities.filter { activity ->
                winners[FamousClimbActivityKey(group, activity.id)]?.resultIndex == resultIndex
            }
            result.copy(activities = filteredActivities, isCompleted = filteredActivities.isNotEmpty())
        }
    }
}

File: back-go/domain/badges/badgeset.go
package badges

import (
	"fmt"
	"mystravastats/internal/shared/domain/business"
	"mystravastats/internal/shared/domain/strava"
	"strings"
)

type BadgeSet struct {
	Name   string
	Badges []Badge
}

func NewBadgeSet(name string, badges []Badge) BadgeSet {
	return BadgeSet{
		Name:   name,
		Badges: badges,
	}
}

func (b BadgeSet) Check(activities []*strava.Activity) []business.BadgeCheckResult {
	var results []business.BadgeCheckResult
	for _, badge := range b.Badges {
		checkedActivities, isCompleted := badge.Check(activities)
		results = append(results, business.BadgeCheckResult{
			Badge:       badge,
			Activities:  checkedActivities,
			IsCompleted: isCompleted,
		})
	}
	return deduplicateFamousClimbActivities(results)
}

type famousClimbActivityWinner struct {
	resultIndex int
	quality     float64
}

func deduplicateFamousClimbActivities(results []business.BadgeCheckResult) []business.BadgeCheckResult {
	winners := make(map[string]map[*strava.Activity]famousClimbActivityWinner)
	for resultIndex, result := range results {
		badge, ok := result.Badge.(FamousClimbBadge)
		if !ok {
			continue
		}
		groupKey := famousClimbGroupKey(badge)
		if winners[groupKey] == nil {
			winners[groupKey] = make(map[*strava.Activity]famousClimbActivityWinner)
		}
		for _, activity := range result.Activities {
			quality, matched := badge.matchQuality(activity)
			if !matched {
				continue
			}
			winner, alreadyAssigned := winners[groupKey][activity]
			if !alreadyAssigned || quality < winner.quality {
				winners[groupKey][activity] = famousClimbActivityWinner{resultIndex: resultIndex, quality: quality}
			}
		}
	}

	for resultIndex := range results {
		badge, ok := results[resultIndex].Badge.(FamousClimbBadge)
		if !ok {
			continue
		}
		groupKey := famousClimbGroupKey(badge)
		filteredActivities := make([]*strava.Activity, 0, len(results[resultIndex].Activities))
		for _, activity := range results[resultIndex].Activities {
			if winner, assigned := winners[groupKey][activity]; assigned && winner.resultIndex == resultIndex {
				filteredActivities = append(filteredActivities, activity)
			}
		}
		results[resultIndex].Activities = filteredActivities
		results[resultIndex].IsCompleted = len(filteredActivities) > 0
	}

	return results
}

func famousClimbGroupKey(badge FamousClimbBadge) string {
	return fmt.Sprintf(
		"%s|%.5f|%.5f",
		strings.ToLower(strings.TrimSpace(badge.Name)),
		badge.End.Latitude,
		badge.End.Longitude,
	)
}

func (b BadgeSet) Plus(anotherBadgeSet BadgeSet) BadgeSet {
	return BadgeSet{
		Name:   b.Name,
		Badges: append(b.Badges, anotherBadgeSet.Badges...),
	}
}
